### PR TITLE
COSBETA-16 Add uploading images to manual journey

### PR DESCRIPTION
**Pull request by DWRendell**

### CODE REVIEW

File: cosmetics-web/app/models/image_upload.rb [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+
+private
+
+  def attached_file_is_image?
+    unless file.attachment.nil? || ALLOWED_CONTENT_TYPES.include?(file.blob.content_type)
+      errors.add :file, "must be one of " + ALLOWED_CONTENT_TYPES.join(", ")
+    end
+  end
+
+  def attached_file_is_within_allowed_size?
```
Comment (DWRendell) [inline cosmetics-web/app/models/image_upload.rb]: I haven't found a good way to test this method without uploading a >30MB file as a fixture (or larger if the size changes, I chose 30MB arbitrarily because it was easyish to test).

Any ideas please let me know

File: cosmetics-web/app/views/notification_build/add_product_image.html.slim [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+        p This can be an artwork file or a photo of the label.
+        p If you upload a photo text in the image must be legible.
+        p The file must be a ‘jpg’, ‘pdf’ ‘png’, or ‘svg’ and no larger than 1gb.
+        - if @notification.errors.messages[:image_uploads].any?
+            span.govuk-error-message Please upload at least one image
+        = form_with(url: wizard_path, method: :put, multipart: true) do |form|
+            .govuk-form-group
+                = form.label :image_upload, "Choose file", class: "govuk-label"
+                = form.file_field :image_upload, multiple: true, class: "govuk-file-upload", \
+                    accept: "image/jpeg, application/pdf, image/png, image/svg+xml"
```
Comment (broder) [inline cosmetics-web/app/views/notification_build/add_product_image.html.slim]: Should we generate this string from `ALLOWED_CONTENT_TYPES`?

File: cosmetics-web/app/views/notifications/edit.html.slim [hunk 3]
```diff
@@ -5,6 +5,7 @@
     .govuk-grid-column-two-thirds
         h1.govuk-heading-xl
             | Check the cosmetic product details before registering
+        = render 'error_summary', errors: @notification.errors
```
Comment (broder) [inline cosmetics-web/app/views/notifications/edit.html.slim]: Are we mixing single quotes and double quotes?

File: cosmetics-web/app/views/notifications/edit.html.slim [hunk 4]
```diff
@@ -60,6 +61,23 @@
...
+                    th.govuk-table__header
+                        | Label image
+                    td.govuk-table__cell
+                        - @notification.image_uploads.each do |image_upload|
+                            - if image_upload.file_exists?
+                                - if image_upload.marked_as_safe?
+                                    = link_to image_upload.filename, url_for(image_upload.file)
+                                - else
+                                        | Processing image
+                                        = " " + image_upload.filename
```
Comment (broder) [inline cosmetics-web/app/views/notifications/edit.html.slim]: I think you can use `'` (rather than `|`) on the line above to add a trailing space to it.

File: cosmetics-web/app/views/notifications/edit.html.slim [hunk 5]
```diff
@@ -60,6 +61,23 @@
...
+                                        | ...
+                            - else
+                                    | Image
+                                    = " " + image_upload.filename
+                                    |  failed anti-virus scan
+                            .br
```
Comment (broder) [inline cosmetics-web/app/views/notifications/edit.html.slim]: Is the `.` here a typo?

File: cosmetics-web/spec/controllers/notification_build_controller_spec.rb [hunk 6]
```diff
@@ -1,6 +1,10 @@
 require 'rails_helper'
 
 RSpec.describe NotificationBuildController, type: :controller do
+  let(:notification) { create(:notification) }
```
Comment (broder) [inline cosmetics-web/spec/controllers/notification_build_controller_spec.rb]: Nice!

File: cosmetics-web/spec/controllers/notifications_controller_spec.rb [hunk 7]
```diff
@@ -29,6 +32,19 @@
       get(:edit, params: { id: notification.id })
       expect(assigns(:notification)).to eq(notification)
     end
+
+    it "adds error if failed attempt to submit when images are pending anti virus check" do
```
Comment (broder) [inline cosmetics-web/spec/controllers/notifications_controller_spec.rb]: I think @aidayounesi is writing similar tests for the zip file parsing - might be worth chatting about different ways of doing it?

File: cosmetics-web/spec/models/notification_spec.rb [hunk 8]
```diff
@@ -6,6 +6,8 @@
     notification.stub(:country_from_code)
       .with('country:NZ')
       .and_return('New Zealand')
+
+    @image_upload = ImageUpload.create
```
Comment (broder) [inline cosmetics-web/spec/models/notification_spec.rb]: Is this being used?

File: cosmetics-web/spec/system/manual_notification_entry_spec.rb [hunk 9]
```diff
@@ -124,13 +154,23 @@ def expect_check_your_answers_value(attribute_name, value)
...
-    else
-      throw "Page URL does not match /notifications/:id/confirmation"
     end
 
     Notification.find(notification_id)
   end
+
+  # The worker doesn't mark system test images as safe, so we have to do it
+  # manually to allow the manual journey to finish.
+  def mark_images_as_safe
```
Comment (broder) [inline cosmetics-web/spec/system/manual_notification_entry_spec.rb]: @aidayounesi is also looking this exact problem!

File: cosmetics-web/app/controllers/notifications_controller.rb [hunk 10]
```diff
@@ -8,14 +8,24 @@ def new
   end
 
     # Check your answers page
-  def edit; end
+  def edit
+    if params[:submit_failed]
```
Comment (broder) [inline cosmetics-web/app/controllers/notifications_controller.rb]: Could we reduce the nesting a bit here? It's a little hard to follow the branching

File: cosmetics-web/app/models/image_upload.rb (hunk 1 above)
Comment (broder) [reply cosmetics-web/app/models/image_upload.rb]: Could we specify the size in a config value and tweak it for the tests? Or allow `attached_file_is_within_allowed_size?` to take a `max_file_size` argument which defaults to `MAX_FILE_BYTES_SIZE`?

Comment (DWRendell) [reply cosmetics-web/app/models/image_upload.rb]: I've gone for reading the filesize from a class method and then stubbing that method for the tests, does that sound sensible?

File: cosmetics-web/app/views/notification_build/add_product_image.html.slim (hunk 2 above)
Comment (DWRendell) [reply cosmetics-web/app/views/notification_build/add_product_image.html.slim]: Yep, done.

File: cosmetics-web/app/views/notifications/edit.html.slim (hunk 3 above)
Comment (DWRendell) [reply cosmetics-web/app/views/notifications/edit.html.slim]: Yes, tried to get rid of singles

File: cosmetics-web/app/views/notifications/edit.html.slim (hunk 4 above)
Comment (DWRendell) [reply cosmetics-web/app/views/notifications/edit.html.slim]: Nice, thanks

File: cosmetics-web/spec/controllers/notifications_controller_spec.rb (hunk 7 above)
Comment (DWRendell) [reply cosmetics-web/spec/controllers/notifications_controller_spec.rb]: Just seen her review, looks nice

File: cosmetics-web/spec/models/notification_spec.rb (hunk 8 above)
Comment (DWRendell) [reply cosmetics-web/spec/models/notification_spec.rb]: No, removed

File: cosmetics-web/app/controllers/notifications_controller.rb (hunk 10 above)
Comment (DWRendell) [reply cosmetics-web/app/controllers/notifications_controller.rb]: Hopefully this is a bit clearer

File: cosmetics-web/app/controllers/notifications_controller.rb [hunk 11]
```diff
@@ -24,4 +28,14 @@ def confirmation
...
     @notification = Notification.find(params[:id])
   end
+
+  def add_image_upload_errors
+    if @notification.images_failed_anti_virus_check?
+      @notification.errors.add :image_uploads, "failed anti virus check"
+    end
+
+    if @notification.images_pending_anti_virus_check?
+      @notification.errors.add :image_uploads, "waiting for files to pass anti virus check..."
```
[the remark this answers is not in the document]
Comment (DWRendell) [reply cosmetics-web/app/controllers/notifications_controller.rb]: Good idea.

File: cosmetics-web/app/models/image_upload.rb [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+    file.attachment.present?
+  end
+
+  def marked_as_safe?
+    file_exists? && file.metadata["safe"]
+  end
+
+  ALLOWED_CONTENT_TYPES = %w[image/jpeg application/pdf image/png image/svg+xml].freeze
+  # 30 MB in Bytes
+  MAX_FILE_BYTES_SIZE = 30000000
```
[the remark this answers is not in the document]
Comment (DWRendell) [reply cosmetics-web/app/models/image_upload.rb]: That's a nice little helper, thanks

File: cosmetics-web/app/models/image_upload.rb [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+
+  validate :attached_file_is_image?
+  validate :attached_file_is_within_allowed_size?
+
+  def file_exists?
+    file.attachment.present?
+  end
+
+  def file_missing?
+    ! file_exists?
```
[the remark this answers is not in the document]
Comment (DWRendell) [reply cosmetics-web/app/models/image_upload.rb]: Thanks, removed

File: cosmetics-web/app/views/notification_build/add_product_image.html.slim [hunk 14]
```diff
@@ -0,0 +1,19 @@
...
+- content_for :after_header do
+    = link_to "Back", previous_wizard_path, class: "govuk-back-link"
+.govuk-grid-row
+    .govuk-grid-column-two-thirds
+        h1.govuk-heading-xl
+            | Upload an image of the product label
+        = render 'error_summary', errors: @notification.errors
+        p This can be an artwork file or a photo of the label.
+        p If you upload a photo text in the image must be legible.
+        p The file must be a ‘jpg’, ‘pdf’ ‘png’, or ‘svg’ and no larger than 1gb.
```
[the remark this answers is not in the document]
Comment (DWRendell) [reply cosmetics-web/app/views/notification_build/add_product_image.html.slim]: 👍

File: cosmetics-web/spec/models/image_upload_spec.rb [hunk 15]
```diff
@@ -0,0 +1,18 @@
+require 'rails_helper'
+
+RSpec.describe ImageUpload, type: :model do
+  let(:image_upload) { ImageUpload.create }
+  let(:file) { fixture_file_upload("testImage.png", "image/png") }
+
+  before do
+    allow(ImageUpload).to receive(:get_max_file_size).and_return(10)
```
[the remark this answers is not in the document]
Comment (DWRendell) [reply cosmetics-web/spec/models/image_upload_spec.rb]: Not sure either, added it to be safe

File: cosmetics-web/app/models/image_upload.rb (hunk 12 above)
[the remark this answers is not in the document]
Comment (DWRendell) [reply cosmetics-web/app/models/image_upload.rb]: Checking for the docs on this I found there's also a `fortnights` method, which I'm sure gets a lot of use

File: cosmetics-web/app/models/image_upload.rb (hunk 1 above)
Comment (broder) [reply cosmetics-web/app/models/image_upload.rb]: Sounds good, yeah
